fame/core/file.py

```diff
@@ -11,14 +11,58 @@
 from fame.core.config import Config
 
 
+def _hash_by_length(hash):
+    _map = {
+        # hashlength: (md5, sha1, sha256)
+        32: (hash, "", ""),
+        40: ("", hash, ""),
+        64: ("", "", hash),
+    }
+
+    return _map.get(len(hash), (None, None, None))
+
+
 class File(MongoDict):
     collection_name = 'files'
 
-    def __init__(self, values=None, filename=None, stream=None, create=True):
+    def __init__(self, values=None, filename=None, stream=None, create=True,
+                 hash=""):
         # When only passing a dict
         if isinstance(values, dict):
+            self['comments'] = []
             MongoDict.__init__(self, values)
+
+        elif hash:
+            MongoDict.__init__(self, {})
+            self['probable_names'] = []
+            self['parent_analyses'] = []
+            self['groups'] = []
+            self['owners'] = []
             self['comments'] = []
+
+            md5, sha1, sha256 = _hash_by_length(hash)
+
+            self.existing = False
+
+            existing_file = (
+                (
+                    # search for existing samples first
+                    self.collection.find_one({'sha256': sha256}) if sha256 else
+                    self.collection.find_one({'sha1': sha1}) if sha1 else
+                    self.collection.find_one({'md5': md5}) if md5 else None
+                )
+                # otherwise, try the hash as filename (aka hash submission)
+                or self.collection.find_one({'names': [hash]})
+            )
+
+            if existing_file:
+                self.existing = True
+                self.update(existing_file)
+            else:
+                self._compute_default_properties(hash_only=True)
+                self._init_hash(hash)
+                self.save()
```

Comment:
The method is starting to get big. Maybe this could all go in a submethod ?

Reply:
Yup, this is definitely getting big. I'll introduce a submethod for the hash initialization code.